### ISSUE TITLE
Error when running the script 'Multimodal_VQA_Captum_Insights.ipynb' using GPU

### ISSUE THREAD

**Issue by allenguo92** [pytorch/captum#469]
## 🐛 Bug

When I run the notebook file Multimodal_VQA_Captum_Insights.ipynb(under 'tutorials') via jupyter notebook, for the last step, it shows the error message "AttributeError: 'NoneType' object has no attribute 'norm'".
## To Reproduce

1.Download and import relevant project 'pytorch-vqa' and 'pytorch-resnet' and pretrained VQA model '2017-08-04_00.55.19.pth' under my own work folder.
2.Change the project and pretrained model path for relevant variables.
3.Run the notebook step by step and finally click the button 'Fetch' after running the command 'visualizer.render()'.

Error messages:
Captum Insights widget created.
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/anaconda3/lib/python3.7/site-packages/captum/insights/widget/widget.py in _fetch_data(self, change)
     39         with self.out:
     40             self.visualizer._update_config(self.config)
---> 41             self.output = namedtuple_to_dict(self.visualizer.visualize())
     42             self.config = dict()
     43 

~/anaconda3/lib/python3.7/site-packages/captum/insights/api.py in visualize(self)
    526         while len(self._outputs) < self._config.num_examples:
    527             try:
--> 528                 self._outputs.extend(self._get_outputs())
    529             except StopIteration:
    530                 break

~/anaconda3/lib/python3.7/site-packages/captum/insights/api.py in _get_outputs(self)
    515             internal_batch_size=1,  # should be 1 until we have batch label support
    516         ):
--> 517             output = self._calculate_vis_output(inputs, additional_forward_args, label)
    518             if output is not None:
    519                 cache = SampleCache(inputs, additional_forward_args, label)

~/anaconda3/lib/python3.7/site-packages/captum/insights/api.py in _calculate_vis_output(self, inputs, additional_forward_args, label, target)
    434             net,
    435             tuple(transformed_inputs),
--> 436             additional_forward_args=additional_forward_args,
    437         )
    438 

~/anaconda3/lib/python3.7/site-packages/captum/attr/_utils/common.py in _run_forward(forward_func, inputs, target, additional_forward_args)
    501         *(*inputs, *additional_forward_args)
    502         if additional_forward_args is not None
--> 503         else inputs
    504     )
    505     return _select_targets(output, target)

~/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py in _call_impl(self, *input, **kwargs)
    720             result = self._slow_forward(*input, **kwargs)
    721         else:
--> 722             result = self.forward(*input, **kwargs)
    723         for hook in itertools.chain(
    724                 _global_forward_hooks.values(),

~/anaconda3/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py in forward(self, *inputs, **kwargs)
    153             return self.module(*inputs[0], **kwargs[0])
    154         replicas = self.replicate(self.module, self.device_ids[:len(inputs)])
--> 155         outputs = self.parallel_apply(replicas, inputs, kwargs)
    156         return self.gather(outputs, self.output_device)
    157 

~/anaconda3/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py in parallel_apply(self, replicas, inputs, kwargs)
    163 
    164     def parallel_apply(self, replicas, inputs, kwargs):
--> 165         return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
    166 
    167     def gather(self, outputs, output_device):

~/anaconda3/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py in parallel_apply(modules, inputs, kwargs_tup, devices)
     83         output = results[i]
     84         if isinstance(output, ExceptionWrapper):
---> 85             output.reraise()
     86         outputs.append(output)
     87     return outputs

~/anaconda3/lib/python3.7/site-packages/torch/_utils.py in reraise(self)
    393             # (https://bugs.python.org/issue2651), so we work around it.
    394             msg = KeyErrorMessage(msg)
--> 395         raise self.exc_type(msg)

AttributeError: Caught AttributeError in replica 0 on device 0.
Original Traceback (most recent call last):
  File "/root/anaconda3/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py", line 60, in _worker
    output = module(*input, **kwargs)
  File "/root/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<ipython-input-6-8ddf6e9a8cf5>", line 27, in forward
    v = v / (v.norm(p=2, dim=1, keepdim=True).expand_as(v) + 1e-8)
AttributeError: 'NoneType' object has no attribute 'norm'

## Expected behavior

![image](https://user-images.githubusercontent.com/4317838/93467275-1c763500-f920-11ea-9e61-eb069eaf69dc.png)
It comes from the tutorials.
## Environment
Describe the environment used for Captum

```

 - Captum / PyTorch Version (e.g., 1.0 / 0.4.0): Captum Version:0.2.0; Pytorch Version: 1.6.0
 - OS (e.g., Linux): Linux 3a6dc91e2776 4.15.0-99-generic #100-Ubuntu SMP Wed Apr 22 20:32:56 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
 - How you installed Captum / PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.7.4
 - CUDA/cuDNN version: V10.2.89
 - GPU models and configuration: 
 - Any other relevant information:

## Additional context

<!-- Add any other context about the problem here. -->

**Comment by bilalsal** [comment]
Hi @allenguo92 

you might be running into a concurrency issue in ResNetLayer4 (cell 7 of the notebook).
The hook defined in __init__ seems not to kick in, before forward() is called.

This might help fix this issue:

import threading
class ResNetLayer4(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.r_model = resnet.resnet152(pretrained=True)
        self.r_model.eval()
        self.r_model.to(device)
        self.buffer = None
        lock = threading.Lock()

        def save_output(module, input, output):
            with lock:
                self.buffer = output

Let us know if you still face the issue (in that case, maybe with details about your GPU setup).

**Comment by allenguo92** [comment]
Hi @bilalsal 

I do remain all the content of the file 'tutorials/Multimodal_VQA_Captum_Insights.ipynb' and then change the corresponding content according to what you implied above. 

However, it still shows the same error log after clicking the 'Fetch' button in the end,


The error log is:

`Captum Insights widget created.
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/anaconda3/lib/python3.7/site-packages/captum/insights/widget/widget.py in _fetch_data(self, change)
     39         with self.out:
     40             self.visualizer._update_config(self.config)
---> 41             self.output = namedtuple_to_dict(self.visualizer.visualize())
     42             self.config = dict()
     43 

~/anaconda3/lib/python3.7/site-packages/captum/insights/api.py in visualize(self)
    526         while len(self._outputs) < self._config.num_examples:
    527             try:
--> 528                 self._outputs.extend(self._get_outputs())
    529             except StopIteration:
    530                 break

~/anaconda3/lib/python3.7/site-packages/captum/insights/api.py in _get_outputs(self)
    515             internal_batch_size=1,  # should be 1 until we have batch label support
    516         ):
--> 517             output = self._calculate_vis_output(inputs, additional_forward_args, label)
    518             if output is not None:
    519                 cache = SampleCache(inputs, additional_forward_args, label)

~/anaconda3/lib/python3.7/site-packages/captum/insights/api.py in _calculate_vis_output(self, inputs, additional_forward_args, label, target)
    434             net,
    435             tuple(transformed_inputs),
--> 436             additional_forward_args=additional_forward_args,
    437         )
    438 

~/anaconda3/lib/python3.7/site-packages/captum/attr/_utils/common.py in _run_forward(forward_func, inputs, target, additional_forward_args)
    501         *(*inputs, *additional_forward_args)
    502         if additional_forward_args is not None
--> 503         else inputs
    504     )
    505     return _select_targets(output, target)

~/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py in _call_impl(self, *input, **kwargs)
    720             result = self._slow_forward(*input, **kwargs)
    721         else:
--> 722             result = self.forward(*input, **kwargs)
    723         for hook in itertools.chain(
    724                 _global_forward_hooks.values(),

~/anaconda3/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py in forward(self, *inputs, **kwargs)
    153             return self.module(*inputs[0], **kwargs[0])
    154         replicas = self.replicate(self.module, self.device_ids[:len(inputs)])
--> 155         outputs = self.parallel_apply(replicas, inputs, kwargs)
    156         return self.gather(outputs, self.output_device)
    157 

~/anaconda3/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py in parallel_apply(self, replicas, inputs, kwargs)
    163 
    164     def parallel_apply(self, replicas, inputs, kwargs):
--> 165         return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
    166 
    167     def gather(self, outputs, output_device):

~/anaconda3/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py in parallel_apply(modules, inputs, kwargs_tup, devices)
     83         output = results[i]
     84         if isinstance(output, ExceptionWrapper):
---> 85             output.reraise()
     86         outputs.append(output)
     87     return outputs

~/anaconda3/lib/python3.7/site-packages/torch/_utils.py in reraise(self)
    393             # (https://bugs.python.org/issue2651), so we work around it.
    394             msg = KeyErrorMessage(msg)
--> 395         raise self.exc_type(msg)

AttributeError: Caught AttributeError in replica 0 on device 0.
Original Traceback (most recent call last):
  File "/root/anaconda3/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py", line 60, in _worker
    output = module(*input, **kwargs)
  File "/root/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<ipython-input-6-609c3a81cab4>", line 29, in forward
    v = v / (v.norm(p=2, dim=1, keepdim=True).expand_as(v) + 1e-8)
AttributeError: 'NoneType' object has no attribute 'norm'
`
I traced the variable 'self.buffer', it was still set as None after running 'forward()' with your code.

**Comment by bilalsal** [comment]
Hi @allenguo92

maybe try apply the lock to the forward method as well?

```python
import threading
class ResNetLayer4(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.r_model = resnet.resnet152(pretrained=True)
        self.r_model.eval()
        self.r_model.to(device)
        self.buffer = None
        lock = threading.Lock()

        def save_output(module, input, output):
            with lock:
                self.buffer = output

    def forward(self, x):
            with lock:
                self.r_model(x)
                return self.buffer
```

If the above does not work, my other suggestions would be to:
1- Use a single node: device_ids=[0] instead of device_ids=[0, 1] in cell 4 and cell 6.
2- Avoid DataParallel altogether
vqa_net =Net(num_tokens) in cell 4 and removing the following line in cell 6:
`vqa_resnet = torch.nn.DataParallel(vqa_resnet, device_ids=[0, 1])
`
3- Avoid the forward_hook in cell 6 by manually performing the forward logic in resnet152 as follows:

```python
import threading
class ResNetLayer4(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.r_model = resnet.resnet152(pretrained=True)
        self.r_model.eval()
        self.r_model.to(device)
        self.buffer = None
        lock = threading.Lock()

        def save_output(module, input, output):
            with lock:
                self.buffer = output

    def forward(self, x):
        x = self.r_model.conv1(x)
        x = self.r_model.bn1(x)
        x = self.r_model.relu(x)
        x = self.r_model.maxpool(x)

        x = self.r_model.layer1(x)
        x = self.r_model.layer2(x)
        x = self.r_model.layer3(x)
        return self.r_model.layer4(x)
```

Hope this helps you further. Let us know if you still face issues.

**Comment by allenguo92** [comment]
Hi @bilalsal ,

I have solved the problem via following your second solution(NOT using forward_hook) and change multi-gpus into one gpu. Also, the code in cell 11 needs to be changed to unify mixed devices into one device:

 img = transform(img).unsqueeze(0)  ---->  img = transform(img).unsqueeze(0).to(device)

For your first solution with lock mechanism, the thread would be blocked after clicking 'Fetch'. 

Anyway, thank you for your kind help.

**Comment by NarineK** [comment]
Thank you @bilalsal  and @allenguo92 for working on this issue. I've upgraded to torch 1.6 and CUDA 10.1 and it works for me.
Couple questions @allenguo92:
1. Have you tried this tutorial? Did it fail too ? 
   https://captum.ai/tutorials/Multimodal_VQA_Interpret
2. Could you, please, tell us your nvidia driver version when you look it up with the following command line?
cat /proc/driver/nvidia/version
There might be some inconsistency between pytorch and the driver version.

**Comment by allenguo92** [comment]
Hi @NarineK 

For your two questions,

1. Yes, I tried, This tutorial works well.
2. The nvidia driver is 'NVRM version: NVIDIA UNIX x86_64 Kernel Module  440.64.00'. The cuda I use is 'Cuda compilation tools, release 10.2, V10.2.89'

**Comment by NarineK** [comment]
@allenguo92, The conda version and the packages look correct. The original problem occurred because I originally made bug fix in: https://captum.ai/tutorials/Multimodal_VQA_Interpret but forgot to make the same fix in the insights.
The hooks should work fine  as it does in `Multimodal_VQA_Interpret`. 
You can still use the same code and just copy the fix that I did long time ago it `Multimodal_VQA_Interpret`, 
basically instead of using `self.buffer = None`, we need to use `self.buffer = {}`
and assign the value as `self.buffer[output.device] = output` instead of `self.buffer = output`.
